---
title: "BRCA-Subtypes"
output: 
  html_document: 
    toc: true
    toc_float:
      collapsed: false
    fig_width: 15
    fig_height: 6
---

```{r warning=FALSE, echo=FALSE, message=FALSE}
## Load package
library(magrittr)
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(ggplot2)
library(patchwork)
library(tidyverse)
library(Matrix)
library(harmony)
library(openxlsx)
library(forestplot)
```

### Import data

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =8, fig.height =6}
load("SC_BRCA_Subtype_prediction.RData")
```

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
SC1 <- subset(x = SC, subset= (FDR < 0.01))
table(SC1$prediction)
table(SC1$subtype,SC1$prediction)
table(SC1$celltype_subset,SC1$prediction)
table(SC1$celltype_major,SC1$prediction)
# table(SC1$Cell_subtype,SC1$prediction)
```

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
id <- SC1@meta.data %>% filter(! celltype_subset %in% "Cancer Cycling")
SC2 <- subset(x = SC1, subset= (celltype_subset %in% id$celltype_subset))
table(SC2$prediction)
table(SC2$subtype,SC2$prediction)
table(SC2$celltype_subset,SC2$prediction)
table(SC2$celltype_major,SC2$prediction)
# table(SC1$Cell_subtype,SC1$prediction)
```

### Heatmap

```{r warning=FALSE, echo=FALSE, fig.width =7, fig.height =5}
tmp <- subset(x = SC2, subset= (celltype_subset %in% 
                                   c("Cancer Basal SC","Cancer Her2 SC",
                                     "Cancer LumA SC","Cancer LumB SC")))
tmp1 <- tmp@meta.data
tmp1$Subtype <- ifelse(tmp1$celltype_subset =="Cancer Basal SC","Basal",
                       ifelse(tmp1$celltype_subset =="Cancer Her2 SC","Her2",
                              ifelse(tmp1$celltype_subset =="Cancer LumA SC","LumA","LumB")))
tmp1$Subtype <- as.factor(tmp1$Subtype)

library(caret)
confusionMatrix(tmp1$Subtype,tmp1$prediction)
```

```{r warning=FALSE, echo=FALSE, fig.width =6, fig.height =6.5}
subt1 <- as.numeric(tmp1$Subtype)
subt2 <- as.numeric(tmp1$prediction)
# calculate consistency
comb.subt <- data.frame(subt1 = paste0("Class", subt1),
                        subt2 = paste0("Class", subt2),
                        stringsAsFactors = F)
tab_classify <- as.data.frame.array(table(comb.subt$subt1,comb.subt$subt2))

# calculate kappa statistic
x <- table(comb.subt$subt2,comb.subt$subt1)
nr <- nrow(x); nc <- ncol(x); N <- sum(x)
Po <- sum(diag(x))/N; Pe <- sum(rowSums(x) * colSums(x)/N)/N
kappa <- (Po - Pe)/(1 - Pe)
seK0 <- sqrt(Pe/(N * (1 - Pe)))
p.v <- 1 - pnorm(kappa/seK0)
p.lab <- ifelse(p.v < 0.001, "P < 0.001", paste0("P = ", format(round(p.v,3), nsmall = 3)))
accuracy <- 0.8412

# generate consistency table
blue   <- "#31587a"
lblue  <- "#457b9d"
dwhite <- "#a8dadc"
white  <- "#f1faee"

# pdf("./Results/Heatmap_of_consistency.pdf", width = 6, height = 6)

par(bty="n", mgp = c(2,0.5,0), mar = c(4.1,4.1,4.1,2.1),tcl=-.25, font.main=1)
par(xpd=NA)
plot(c(0,ncol(tab_classify)),c(0,nrow(tab_classify)),
     col = "white",cex.main=1.5,line = 2.5,
     xlab = "",xaxt = "n",
     ylab = "",yaxt = "n")
title(paste0("Accuracy = ", format(round(accuracy,2), nsmall = 2),
             "\n", 
             "Kappa = ", format(round(kappa,3), nsmall = 2),
             "\n", p.lab),
      adj = 0, line = 0)

# add y-axis
axis(2, at = 0.5:(nrow(tab_classify)-0.5), labels = FALSE)
text(y = 0.5:(nrow(tab_classify)-0.5),
     par("usr")[1],
     labels = rownames(tab_classify)[nrow(tab_classify):1],
     srt = 0, pos = 2, cex=1, xpd = TRUE)
mtext(paste0("SCSubtype prediction"),cex=1.2,  side = 2, line = 3)

# add x-axis
axis(1, at = 0.5:(ncol(tab_classify)-0.5), labels = FALSE)
text(x = 0.5:(ncol(tab_classify)-0.5),
     par("usr")[1] - 0.1,
     labels = colnames(tab_classify),
     srt = 0, pos = 1,cex=1, xpd = TRUE)
mtext(paste0("PIPET prediction"),cex=1.2, side = 1, line = 2)

# generate colors
input_matrix <- as.matrix(tab_classify)
mat.max = max(input_matrix)
unq.value <- unique(sort(as.vector(input_matrix)))
rbPal <- colorRampPalette(c(white,dwhite,lblue,blue))
col.vec <- rbPal(max(unq.value) + 1)
col.mat <- matrix(NA,byrow = T,ncol = ncol(input_matrix),nrow = nrow(input_matrix))

# fill matrix
for (i in 1:nrow(col.mat)) {
  for (j in 1:ncol(col.mat)) {
    col.mat[i,j] <- col.vec[input_matrix[i,j] + 1]
  }
}

# generate heatmap
x_size <- ncol(input_matrix)
y_size <- nrow(input_matrix)

my_xleft = rep(c(0:(x_size-1)),each = x_size)
my_xright = my_xleft + 1
my_ybottom = rep(c((y_size-1):0),y_size)
my_ytop = my_ybottom + 1
rect(xleft = my_xleft,
     ybottom = my_ybottom,
     xright = my_xright,
     ytop = my_ytop,
     col=col.mat,
     border = F)

# fill count
text(my_xleft + 0.5,my_ybottom + 0.5,input_matrix, cex = 1)

# dev.off()
```

### Pie plots

#### Basal

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
tmp1 <- subset(SC2, subset = prediction %in% c("Basal"))
table(tmp1$celltype_subset,tmp1$prediction)
```

```{r warning=FALSE, echo=FALSE, fig.width =7.5, fig.height =6}
# x <- ifelse(tmp1$Cell_subtype == "tS1","Tumor cell states 1",
#             ifelse(tmp1$Cell_subtype == "tS2","Tumor cell states 2",
#                    ifelse(tmp1$Cell_subtype == "tS3","Tumor cell states 3",tmp1$Cell_subtype)))
x <- tmp1$celltype_subset
tmp0 <- data.frame(table(x))
tmp0$y <- ifelse(tmp0$Freq>10,levels(tmp0$x),"Others")
tmp <- aggregate(x = tmp0$Freq, by= list(tmp0$y), FUN = sum)
colnames(tmp) <- c("x","Freq")
tmp$percent<- paste(round(100*tmp$Freq/sum(tmp$Freq), 2), "%")
tmp <- tmp[order(tmp$Freq,decreasing = TRUE),]
tmp$celltype <- factor(tmp$x,levels=tmp$x)

library(RColorBrewer)
colors<-brewer.pal(12,'Paired')
# colors<-c("#3f60aa","#44c1f0","#13a983","#9ec417",
#           "#e4ce00","#f18800","#e8490f","#cc340c")

library(ggforce)
p1 <- ggplot()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank(), 
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title=element_blank(), 
        panel.border = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size=18,hjust = 0.5,face="bold"))+
  xlab("")+ylab('')+ ggtitle("Basal cells") +
  scale_fill_manual(values = colors)+
  geom_arc_bar(data=tmp,
               stat = "pie",
               aes(x0=0,y0=0,r0=1,r=2,
                   amount=Freq,fill=celltype)
  )+
  annotate("text",x=1.4,y=-1.6,label="77.06%",angle=42,size=5)+
  annotate("text",x=-1.6,y=1.35,label="17.58%",angle=55,size=5)
p1

# pdf("./Results/BRCA_Basal_Pie_plot.pdf",width=7.5,height=6)
# p1
# dev.off()
```

#### Her2

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
tmp2 <- subset(SC2, subset = prediction %in% c("Her2"))
table(tmp2$celltype_subset,tmp2$prediction)
```

```{r warning=FALSE, echo=FALSE, fig.width =7.5, fig.height =6}
x <- tmp2$celltype_subset
tmp0 <- data.frame(table(x))
tmp0$y <- ifelse(tmp0$Freq>10,levels(tmp0$x),"Others")
tmp <- aggregate(x = tmp0$Freq, by= list(tmp0$y), FUN = sum)
colnames(tmp) <- c("x","Freq")
tmp$percent<- paste(round(100*tmp$Freq/sum(tmp$Freq), 2), "%")
tmp <- tmp[order(tmp$Freq,decreasing = TRUE),]
tmp$celltype <- factor(tmp$x,levels=tmp$x)

library(RColorBrewer)
colors<-brewer.pal(12,'Paired')
# colors<-c("#3f60aa","#44c1f0","#13a983","#9ec417",
#           "#e4ce00","#f18800","#e8490f","#cc340c")

p2 <- ggplot()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank(), 
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title=element_blank(), 
        panel.border = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size=18,hjust = 0.5,face="bold"))+
  xlab("")+ylab('')+ ggtitle("Her2 cells") +
  scale_fill_manual(values = colors)+
  geom_arc_bar(data=tmp,
               stat = "pie",
               aes(x0=0,y0=0,r0=1,r=2,
                   amount=Freq,fill=celltype)
  )+
  annotate("text",x=1.85,y=-1.05,label="66.17%",angle=55,size=5)+
  annotate("text",x=-2.12,y=-0.1,label="16.35%",angle=95,size=5)+
  annotate("text",x=-1.5,y=1.5,label="9.10%",angle=50,size=5)+
  annotate("text",x=-0.75,y=2,label="4.76%",angle=20,size=5)
p2

# pdf("./Results/BRCA_Her2_Pie_plot.pdf",width=7.5,height=6)
# p2
# dev.off()
```

#### LumA

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
tmp3 <- subset(SC2, subset = prediction %in% c("LumA"))
table(tmp3$celltype_subset,tmp3$prediction)
```

```{r warning=FALSE, echo=FALSE, fig.width =7.5, fig.height =6}
x <- tmp3$celltype_subset
tmp0 <- data.frame(table(x))
tmp0$y <- ifelse(tmp0$Freq>10,levels(tmp0$x),"Others")
tmp <- aggregate(x = tmp0$Freq, by= list(tmp0$y), FUN = sum)
colnames(tmp) <- c("x","Freq")
tmp$percent<- paste(round(100*tmp$Freq/sum(tmp$Freq), 2), "%")
tmp <- tmp[order(tmp$Freq,decreasing = TRUE),]
tmp$celltype <- factor(tmp$x,levels=tmp$x)

p3 <- ggplot()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank(), 
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title=element_blank(), 
        panel.border = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size=18,hjust = 0.5,face="bold"))+
  xlab("")+ylab('')+ ggtitle("LumA cells") +
  scale_fill_manual(values = colors)+
  geom_arc_bar(data=tmp,
               stat = "pie",
               aes(x0=0,y0=0,r0=1,r=2,
                   amount=Freq,fill=celltype)
  )+
  annotate("text",x=1.4,y=-1.6,label="76.01%",angle=42,size=5)+
  annotate("text",x=-1.9,y=0.95,label="11.82%",angle=60,size=5)+
  annotate("text",x=-1.3,y=1.67,label="3.91%",angle=35,size=5) +
  annotate("text",x=-0.85,y=1.92,label="3.45%",angle=25,size=5) 
p3

# pdf("./Results/BRCA_LumA_Pie_plot.pdf",width=7.5,height=6)
# p3
# dev.off()
```

#### LumB

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
tmp4 <- subset(SC2, subset = prediction %in% c("LumB"))
table(tmp4$celltype_subset,tmp4$prediction)
```

```{r warning=FALSE, echo=FALSE, fig.width =7.5, fig.height =6}
x <- tmp4$celltype_subset
tmp0 <- data.frame(table(x))
tmp0$y <- ifelse(tmp0$Freq>10,levels(tmp0$x),"Others")
tmp <- aggregate(x = tmp0$Freq, by= list(tmp0$y), FUN = sum)
colnames(tmp) <- c("x","Freq")
tmp$percent<- paste(round(100*tmp$Freq/sum(tmp$Freq), 2), "%")
tmp <- tmp[order(tmp$Freq,decreasing = TRUE),]
tmp$celltype <- factor(tmp$x,levels=tmp$x)

library(RColorBrewer)
colors<-brewer.pal(12,'Paired')

library(ggforce)
p4 <- ggplot()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank(), 
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title=element_blank(), 
        panel.border = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size=18,hjust = 0.5,face="bold"))+
  xlab("")+ylab('')+ ggtitle("LumB cells") +
  scale_fill_manual(values = colors)+
  geom_arc_bar(data=tmp,
               stat = "pie",
               aes(x0=0,y0=0,r0=1,r=2,
                   amount=Freq,fill=celltype)
  )+
  annotate("text",x=1.75,y=-1.2,label="72.82%",angle=50,size=5)+
  annotate("text",x=-2.08,y=0.4,label="11.33%",angle=82,size=5)+
  annotate("text",x=-1.6,y=1.36,label="4.56%",angle=48,size=5) +
  annotate("text",x=-1.20,y=1.72,label="2.43%",angle=35,size=5) 
p4

# pdf("./Results/BRCA_LumB_Pie_plot.pdf",width=7.8,height=6)
# p4
# dev.off()
```

### DE genes

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
DefaultAssay(SC2) <- "RNA"
Idents(SC2)="prediction"

#find markers for every cluster compared to all remaining cells, report only the positive ones
SC.markers <- FindAllMarkers(SC2, only.pos = TRUE, 
                             min.pct = 0.25, logfc.threshold = 0.25)

biomarkers <- SC.markers %>%
  dplyr::filter(p_val_adj<0.05) %>%
  group_by(cluster) %>%
  slice_max(n = 20000, order_by = avg_log2FC)

Basal <- biomarkers %>% dplyr::filter(cluster=="Basal") %>%
  dplyr::filter(avg_log2FC>=2)
Her2 <- biomarkers %>% dplyr::filter(cluster=="Her2") %>%
  dplyr::filter(avg_log2FC>=2)
LumA <- biomarkers %>% dplyr::filter(cluster=="LumA") %>%
  dplyr::filter(avg_log2FC>=2)
LumB <- biomarkers %>% dplyr::filter(cluster=="LumB") %>%
  dplyr::filter(avg_log2FC>=2)

# tmp <- biomarkers %>% dplyr::filter(cluster %in% c("Basal","Her2","LumA","LumB")) %>%
#   dplyr::filter(avg_log2FC>=2)
table <- rbind(Basal,Her2,LumA,LumB)
write.xlsx(table, "./Results/BRCA_Subtype_biomarkers_pos_FDR0.05_orderByavglog2FC.xlsx")
```

### Multi-Violin Plot

```{r warning=FALSE, echo=FALSE, fig.width =15, fig.height =4}
library(cowplot)

features <- c("LDHB","RARRES1","MMP7","CRYAB","KRT5","KRT14","KRT15","ANXA1",
              "CD24","MUCL1","S100A8","S100A9","IDH2","FABP7",
              "SCGB2A2","SCGB1D2","PIP","AZGP1","PHGR1",
              "H2AFJ","SLC39A6","STC2","DHRS2","COX6C")

## Normalizing
SC2 <- NormalizeData(SC2, normalization.method = "LogNormalize", scale.factor = 1e4)
## Find highly variable features
SC2 <- FindVariableFeatures(SC2, selection.method = "vst", nfeatures = 2000)
## Scaling the data
SC2 <- ScaleData(SC2, features = rownames(SC2))
#features <- Mutated$gene
data <- data.frame(t(data.frame(SC2@assays$RNA@data[features,])))

# Add cell ID and identity classes
data$Cell <- colnames(SC2)
data$Celltype <- SC2$prediction

# Use melt to change data.frame format
data1 <- reshape2::melt(data, id.vars = c("Cell","Celltype"), measure.vars = features,
                       variable.name = "Feat", value.name = "Expr")
head(data1)

p_1 <- ggplot(data1, aes(Expr, Celltype, fill = Celltype)) +
        geom_violin(scale = "width", adjust = 1, trim = TRUE) +
        scale_x_continuous(expand = c(0, 0), labels = function(x)
                           c(rep(x = "", times = length(x)-2), x[length(x) - 1], "")) +
        facet_grid(cols = vars(Feat), scales = "free")  +
        theme_cowplot(font_size = 12) +
        scale_fill_brewer(palette = 'Set1')+
        theme(legend.position = "none", 
 #             panel.spacing = unit(0, "lines"),
          plot.title = element_text(hjust = 0.5),
 #             panel.background = element_rect(fill = NA, color = "black"),
              strip.background = element_blank(),
              strip.text = element_text(face = "bold"),
              strip.text.x.top = element_text(angle = 90, hjust = 0, vjust = 0.5)) +
        xlab("Gene Expression") + ylab("")
p_1

# pdf("./Results/BRAC_Subtype_Multi_Violin.pdf",width=15,height=4)
# p_1
# dev.off()
```

### GSEA by using KEGG genesets

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 4, fig.width = 9}
library(fgsea)
library(msigdbr)
library(openxlsx)

mdb_c2 <- msigdbr(species = "Homo sapiens", category = "C2")
mdb_kegg = mdb_c2 [grep("^KEGG",mdb_c2 $gs_name),]
fgsea_sets = mdb_kegg %>% split(x = .$gene_symbol, f = .$gs_name)

biomarkers <- SC.markers %>%
  dplyr::filter(p_val_adj<0.05) 
biomarkers1 <- biomarkers %>%
  dplyr::filter(cluster %in% c("Her2","LumA")) 
biomarkers2 <- biomarkers %>%
  dplyr::filter(cluster %in% c("Basal","LumB")) %>%
  dplyr::filter(avg_log2FC>=0.5)
biomarkers <- rbind(biomarkers1,biomarkers2)
#length(fgsea_sets)

cluster <- c("Basal","Her2","LumA","LumB")
ls <- list()
for (i in cluster) {
  makers <- biomarkers %>% filter(cluster==i)
  cluster.genes<- makers %>% arrange(desc(avg_log2FC)) %>% dplyr::select(gene,avg_log2FC)
  ranks<- deframe(cluster.genes)
  fgseaRes<- fgsea(fgsea_sets, stats = ranks)
  fgseaRes<- fgseaRes %>% arrange(desc(NES))
  ls <- c(ls,list(fgseaRes))
}
names(ls)[1] <- "Basal"
names(ls)[2] <- "Her2"
names(ls)[3] <- "LumA"
names(ls)[4] <- "LumB"
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 5, fig.width = 8}
ls1 <- ls[["Basal"]]
ls1$pathway <- gsub('[_]', ' ', ls1$pathway)
ls1$pathway <- str_replace(ls1$pathway, "KEGG ","")
ls1$pathway <- str_replace(ls1$pathway, "CYTOKINE CYTOKINE ","CYTOKINE-CYTOKINE")

tmp1 <- ls1 %>% as_tibble() %>%
  arrange(desc(NES)) %>%
  filter(pval < 0.05) %>%
  filter(NES > 0)

tmp1$Subtype <- "Basal"
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 4, fig.width = 8}
ls2 <- ls[["Her2"]]
ls2$pathway <- gsub('[_]', ' ', ls2$pathway)
ls2$pathway <- str_replace(ls2$pathway, "KEGG ","")
ls2$pathway <- str_replace(ls2$pathway, "CYTOKINE CYTOKINE ","CYTOKINE-CYTOKINE")

tmp2 <- ls2 %>% as_tibble() %>%
  arrange(desc(NES)) %>%
  filter(pval < 0.05) %>%
  filter(NES > 0)

tmp2$Subtype <- "Her2"
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 4, fig.width = 8}
ls3 <- ls[["LumA"]]
ls3$pathway <- gsub('[_]', ' ', ls3$pathway)
ls3$pathway <- str_replace(ls3$pathway, "KEGG ","")
ls3$pathway <- str_replace(ls3$pathway, "CYTOKINE CYTOKINE ","CYTOKINE-CYTOKINE")

tmp3 <- ls3 %>% as_tibble() %>%
  arrange(desc(NES)) %>%
  filter(pval < 0.05) %>%
  filter(NES > 0)

tmp3$Subtype <- "LumA"
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 6, fig.width = 8}
ls4 <- ls[["LumB"]]
ls4$pathway <- gsub('[_]', ' ', ls4$pathway)
ls4$pathway <- str_replace(ls4$pathway, "KEGG ","")
ls4$pathway <- str_replace(ls4$pathway, "CYTOKINE CYTOKINE ","CYTOKINE-CYTOKINE")

tmp4 <- ls4 %>% as_tibble() %>%
  arrange(desc(NES)) %>%
  filter(pval < 0.05)%>%
  filter(NES > 0)

tmp4$Subtype <- "LumB"
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 8, fig.width = 9}
tmp <- rbind(tmp1,tmp2,tmp3,tmp4)  
tmp$num <- 1:nrow(tmp)

p <- ggplot(tmp, aes(reorder(pathway, -num), NES,fill= Subtype)) +
  geom_bar(stat ="identity",width = 0.8,position = position_dodge(width=0.8))+ 
#  scale_fill_gradient(low = "#156077", high = "#f46f20")+
  scale_fill_brewer(palette = 'Set1')+
  coord_flip() +
  labs(x="")+
  theme_bw()+
  # theme_classic()+
    theme(#panel.grid.major.x = element_line(color = "grey"), 
          panel.grid = element_blank(),
          axis.text.x = element_text(size = 13, color = "black"),
          axis.text.y = element_text(size = 12, color = "black"),
          axis.title.x = element_text(size = 16), 
          axis.title.y = element_text(size = 16),
          legend.text = element_text(size = 12), 
          legend.title = element_blank())
p

# pdf(file.path("./Results/BRCA_Subtype_KEGG_pathway.pdf"),width = 10,height = 8)
# p
# dev.off()
```

### GSEA by using HALLMARK genesets

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 8, fig.width = 9}
library(clusterProfiler)
library(enrichplot)

MSigDB=read.gmt("h.all.v2023.1.Hs.symbols.gmt")
MSigDB$term <- str_replace(MSigDB$term, "HALLMARK_","")
MSigDB$term <- str_replace_all(MSigDB$term, "_"," ")

biomarkers <- SC.markers %>%
  dplyr::filter(p_val_adj<0.05) 

data <- biomarkers
data$log2fc <- ifelse(data$cluster %in% c("Basal"),data$avg_log2FC,-data$avg_log2FC)
geneList <- data$log2fc
names(geneList) <- data$gene
geneList <- sort(geneList,decreasing = T)
GSEA.data <- GSEA(geneList = geneList,TERM2GENE=MSigDB,seed = T,verbose=F,pvalueCutoff = 0.25)
tmp1 <- GSEA.data@result %>% 
  arrange(desc(NES)) %>% 
  filter(NES>0)
tmp1$Subtype <- "Basal"

data <- biomarkers
data$log2fc <- ifelse(data$cluster %in% c("Her2"),data$avg_log2FC,-data$avg_log2FC)
geneList <- data$log2fc
names(geneList) <- data$gene
geneList <- sort(geneList,decreasing = T)
GSEA.data <- GSEA(geneList = geneList,TERM2GENE=MSigDB,seed = T,verbose=F,pvalueCutoff = 0.25)
tmp2 <- GSEA.data@result %>% 
  arrange(desc(NES)) %>% 
  filter(NES>0)
tmp2$Subtype <- "Her2"

data <- biomarkers
data$log2fc <- ifelse(data$cluster %in% c("LumA"),data$avg_log2FC,-data$avg_log2FC)
geneList <- data$log2fc
names(geneList) <- data$gene
geneList <- sort(geneList,decreasing = T)
GSEA.data <- GSEA(geneList = geneList,TERM2GENE=MSigDB,seed = T,verbose=F,pvalueCutoff = 0.25)
tmp3 <- GSEA.data@result %>% 
  arrange(desc(NES)) %>% 
  filter(NES>0)
tmp3$Subtype <- "LumA"

data <- biomarkers
data$log2fc <- ifelse(data$cluster %in% c("LumB"),data$avg_log2FC,-data$avg_log2FC)
geneList <- data$log2fc
names(geneList) <- data$gene
geneList <- sort(geneList,decreasing = T)
GSEA.data <- GSEA(geneList = geneList,TERM2GENE=MSigDB,seed = T,verbose=F,pvalueCutoff = 0.25)
tmp4 <- GSEA.data@result %>% 
  arrange(desc(NES)) %>% 
  filter(NES>0)
tmp4$Subtype <- "LumB"
 
tmp <- rbind(tmp1,tmp2,tmp3,tmp4)  
tmp$num <- 1:nrow(tmp)
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 8, fig.width = 9}
p <- ggplot(tmp, aes(reorder(ID, -num), NES,fill= Subtype)) +
  geom_bar(stat ="identity",width = 0.8,position = position_dodge(width=0.8))+ 
#  scale_fill_gradient(low = "#156077", high = "#f46f20")+
  scale_fill_brewer(palette = 'Set1')+
  coord_flip() +
  labs(x="")+
  theme_bw()+
  # theme_classic()+
    theme(#panel.grid.major.x = element_line(color = "grey"), 
          panel.grid = element_blank(),
          axis.text.x = element_text(size = 13, color = "black"),
          axis.text.y = element_text(size = 12, color = "black"),
          axis.title.x = element_text(size = 16), 
          axis.title.y = element_text(size = 16),
          legend.text = element_text(size = 12), 
          legend.title = element_blank())
p

# pdf(file.path("./Results/BRCA_Subtype_GSEA_pathway.pdf"),width = 9,height = 8)
# p
# dev.off()
```
