---
title: "LUAD-Survival"
output: 
  html_document: 
    toc: true
    toc_float:
      collapsed: false
    fig_width: 15
    fig_height: 6
---

```{r warning=FALSE, echo=FALSE, message=FALSE}
## Load package
library(magrittr)
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(ggplot2)
library(patchwork)
library(tidyverse)
library(Matrix)
library(harmony)
library(openxlsx)
library(forestplot)
```

### Import data

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =8, fig.height =6}
load("SC_LUAD_surv_prediction.RData")
```

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
SC1 <- subset(x = SC_Lung, subset= (Pvalue < 0.05))
table(SC1$Cell_type,SC1$prediction)
# table(SC1$Cell_subtype,SC1$prediction)
```

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
tmp <- subset(SC1, subset = prediction %in% c("Died"))
tmp1 <- subset(tmp, subset = Cell_type %in% c("Epithelial cells"))
tmp2 <- subset(tmp, subset = Cell_type %in% c("Myeloid cells"))

table(tmp1$Cell_subtype,tmp1$prediction)
table(tmp2$Cell_subtype,tmp2$prediction)

x <- ifelse(tmp1$Cell_subtype == "tS1","Tumor cell states 1",
            ifelse(tmp1$Cell_subtype == "tS2","Tumor cell states 2",
                   ifelse(tmp1$Cell_subtype == "tS3","Tumor cell states 3",tmp1$Cell_subtype)))
tmp <- data.frame(table(x))
tmp$percent<- paste(round(100*tmp$Freq/sum(tmp$Freq), 2), "%")
tmp <- tmp[order(tmp$Freq,decreasing = TRUE),]
tmp$celltype <- factor(tmp$x,levels=tmp$x)

library(RColorBrewer)
colors<-brewer.pal(8,'Paired')
# colors<-c("#3f60aa","#44c1f0","#13a983","#9ec417",
#           "#e4ce00","#f18800","#e8490f","#cc340c")

library(ggforce)
p1 <- ggplot()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank(), 
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title=element_blank(), 
        panel.border = element_blank(),
        panel.background = element_blank())+
  xlab("")+ylab('')+
  scale_fill_manual(values = colors)+
  geom_arc_bar(data=tmp,
               stat = "pie",
               aes(x0=0,y0=0,r0=1,r=2,
                   amount=Freq,fill=celltype)
  )+
  annotate("text",x=2.1,y=-0.5,label="57.58%",angle=80,size=5)+
  annotate("text",x=-2.2,y=0,label="34.47%",angle=90,size=5)+
  annotate("text",x=-0.6,y=2.05,label="6.46%",angle=10,size=5)
p1
```

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
x <- tmp2$Cell_subtype
tmp <- data.frame(table(x))
tmp$percent<- paste(round(100*tmp$Freq/sum(tmp$Freq), 2), "%")
tmp <- tmp[order(tmp$Freq,decreasing = TRUE),]
tmp$celltype <- factor(tmp$x,levels=tmp$x)

colors<-brewer.pal(12,'Paired')
# colors<-c("#3f60aa","#44c1f0","#13a983","#9ec417",
#           "#e4ce00","#f18800","#e8490f","#cc340c")

p2 <- ggplot()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank(), 
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title=element_blank(), 
        panel.border = element_blank(),
        panel.background = element_blank())+
  xlab("")+ylab('')+
  scale_fill_manual(values = colors)+
  geom_arc_bar(data=tmp,
               stat = "pie",
               aes(x0=0,y0=0,r0=1,r=2,
                   amount=Freq,fill=celltype)
  )+
  annotate("text",x=2.2,y=0.3,label="49.93%",angle=95,size=5)+
  annotate("text",x=-1.35,y=-1.8,label="20.52%",angle=-45,size=5)+
  annotate("text",x=-2.2,y=0.2,label="6.57%",angle=85,size=5)+
  annotate("text",x=-1.8,y=1.3,label="11.76%",angle=55,size=5)+
  annotate("text",x=-1.1,y=1.9,label="5.61%",angle=35,size=5)
p2
```

```{r warning=FALSE, echo=FALSE, fig.width =12, fig.height =4}
library(ggpubr)
ggarrange(p1,p2,ncol = 2,nrow =1)

pdf("./Results/LUAD_Epithelial_Myeloid_Pie.pdf",width=12,height=4)
ggarrange(p1,p2,ncol = 2,nrow =1)
dev.off()
```

### Proportion plots

```{r warning=FALSE, echo=FALSE, fig.width =7, fig.height =7}
tmp <- as.data.frame(table(SC1$Sample_Origin,SC1$prediction))
tmp <- tmp %>%
  group_by(Var2) %>%
  mutate(prop = round(Freq/sum(Freq)*100, digits = 1))
tmp$tissue <- ifelse(tmp$Var1 %in% c("nLung"),"normal Lung","tumor Lung")

p <- ggplot(tmp, aes(x = tissue, y = prop,fill = Var2))+
    #####这部分的position_dodge(width=0.8)大于宽width = 0.6点，可以使得分组内柱子之间有缝隙，而不是贴合。
    geom_bar(stat ="identity",width = 0.6,position = position_dodge(width=0.8))+        
    scale_fill_manual(values = c("#E7B800","#8ac926"))+
    labs(x = "",y = "Proportion(%)")+
    ###########文字的position设置类似bar的position
    geom_text(aes(label = tmp$prop),position=position_dodge(width = 0.8),size =4 ,vjust = -0.45)+  
    guides(fill = guide_legend(reverse = F))+
    theme_classic()+
    scale_y_continuous(expand=c(0,0))+
    coord_cartesian(ylim = c(0, 100)) +
    theme(axis.text.x = element_text(size = 13, color = "black"),
          axis.text.y = element_text(size = 12, color = "black"),
          axis.title.x = element_text(size = 16), 
          axis.title.y = element_text(size = 16),
          legend.text = element_text(size = 12), 
          legend.title = element_text(size = 0))
p

pdf("./Results/LUAD_Died_Alive_Proportion_of_normal_tumor.pdf",width=7,height=7)
p
dev.off()
```

```{r warning=FALSE, echo=FALSE, fig.width =5, fig.height =7}
tmp <- as.data.frame(table(SC1$Cell_type,SC1$prediction))
tmp <- tmp %>%
  filter(Var2=="Died") %>%
  mutate(prop = round(Freq/sum(Freq)*100, digits = 1))

p <- ggplot(tmp, aes(x = reorder(Var1, - Freq), y = Freq, fill = Var1)) +
  geom_bar(stat ="identity",width = 0.6,position = position_dodge(width=0.8))+        
#    scale_fill_manual(values = c("#E7B800","#8ac926"))+
    labs(x = "",y = "Number of cells")+
    ###########文字的position设置类似bar的position
    geom_text(aes(label = tmp$Freq),position=position_dodge(width = 0.8),size =4 ,vjust = -0.45)+  
#    guides(fill = guide_legend(reverse = F))+
    theme_classic()+
    scale_y_continuous(expand=c(0,0))+
    coord_cartesian(ylim = c(0, 2000)) +
    theme(axis.text.x = element_text(size = 13, color = "black",angle = 90,hjust = 1,vjust = 0.3),
          axis.text.y = element_text(size = 12, color = "black"),
          axis.title.x = element_text(size = 16), 
          axis.title.y = element_text(size = 16),
          legend.text = element_text(size = 12), 
          legend.position="none")
p 

pdf("./Results/LUAD_Died_Cells_in_Celltypes.pdf",width=6,height=7)
p
dev.off()
```

### DE genes

```{r warning=FALSE, echo=FALSE, fig.width =8, fig.height =6}
DefaultAssay(SC1) <- "RNA"
Idents(SC1)="prediction"

#find markers for every cluster compared to all remaining cells, report only the positive ones
SC.markers <- FindAllMarkers(SC1, only.pos = TRUE, 
                             min.pct = 0.25, logfc.threshold = 0.25)

biomarkers <- SC.markers %>%
  dplyr::filter(p_val_adj<0.05) %>%
  group_by(cluster) %>%
  slice_max(n = 20000, order_by = avg_log2FC)

Died <- biomarkers %>% dplyr::filter(cluster=="Died") %>%
  dplyr::filter(avg_log2FC>=2)

tmp <- biomarkers %>% dplyr::filter(cluster %in% c("Died","Alive")) %>%
  dplyr::filter(avg_log2FC>=2)
```

```{r warning=FALSE, echo=FALSE, fig.width =12, fig.height =8}
# Subset data.frame
## Normalizing
SC1 <- NormalizeData(SC1, normalization.method = "LogNormalize", scale.factor = 1e4)
## Find highly variable features
SC1 <- FindVariableFeatures(SC1, selection.method = "vst", nfeatures = 2000)
## Scaling the data
SC1 <- ScaleData(SC1, features = rownames(SC1))

library(ggsignif)

p1 <- VlnPlot(SC1,features = "NAPSA",adjust = .8,
        cols = c("#E7B800","#8ac926"), group.by ="prediction",pt.size = 0)+
  xlab("") + stat_compare_means()
p2 <- VlnPlot(SC1,features = "COL1A1",adjust = .8,
        cols = c("#E7B800","#8ac926"), group.by ="prediction",pt.size = 0)+
  xlab("") + stat_compare_means()
p3 <- VlnPlot(SC1,features = "TM4SF1",adjust = .8,
        cols = c("#E7B800","#8ac926"), group.by ="prediction",pt.size = 0)+
  xlab("") + stat_compare_means()
p4 <- VlnPlot(SC1,features = "MDK",adjust = .8,
        cols = c("#E7B800","#8ac926"), group.by ="prediction",pt.size = 0)+
  xlab("") + stat_compare_means()
p5 <- VlnPlot(SC1,features = "CEACAM6",adjust = .8,
        cols = c("#E7B800","#8ac926"), group.by ="prediction",pt.size = 0)+
  xlab("") + stat_compare_means()
p6 <- VlnPlot(SC1,features = "KRT19",adjust = .8,
        cols = c("#E7B800","#8ac926"), group.by ="prediction",pt.size = 0)+
  xlab("") + stat_compare_means()

library(ggpubr)
ggarrange(p1,p2,p3,p4,p5,p6,ncol = 3,nrow =2)

pdf("./Results/LUAD_Died_genes.pdf",width=12,height=8)
ggarrange(p1,p2,p3,p4,p5,p6,ncol = 3,nrow =2)
dev.off()
```

### Multi-Violin Plot

```{r warning=FALSE, echo=FALSE, fig.width =10, fig.height =6}
library(cowplot)

features <- Died$gene
data <- data.frame(t(data.frame(SC1@assays$RNA@data[features,])))

# Add cell ID and identity classes
data$Cell <- colnames(SC1)
data$Celltype <- factor(SC1$Cell_type)

# Use melt to change data.frame format
data1 <- reshape2::melt(data, id.vars = c("Cell","Celltype"), measure.vars = features,
                       variable.name = "Feat", value.name = "Expr")
head(data1)

p_1 <- ggplot(data1, aes(Expr, Celltype, fill = Celltype)) +
        geom_violin(scale = "width", adjust = 1, trim = TRUE) +
        scale_x_continuous(expand = c(0, 0), labels = function(x)
                           c(rep(x = "", times = length(x)-2), x[length(x) - 1], "")) +
        facet_grid(cols = vars(Feat), scales = "free")  +
        theme_cowplot(font_size = 12) +
        theme(legend.position = "none", 
 #             panel.spacing = unit(0, "lines"),
          plot.title = element_text(hjust = 0.5),
 #             panel.background = element_rect(fill = NA, color = "black"),
              strip.background = element_blank(),
              strip.text = element_text(face = "bold"),
              strip.text.x.top = element_text(angle = 90, hjust = 0, vjust = 0.5)) +
        xlab("Gene Expression") + ylab("")
p_1
```

### Cox

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =5, fig.height =6}
library('survival')
library('survminer')

TPM.norm <- log2(TPM + 1)
TPM.norm <- data.frame(t(na.omit(TPM.norm[Died$gene,])))

survival_info_df <- surv[,c(1,2)]
tmp <-TPM.norm[rownames(survival_info_df),]
survival_info_df <- cbind(tmp,survival_info_df)

multi_COX<-coxph(Surv(futime, fustat) ~ ., data=survival_info_df)
multi_COX

genes <- c("SFTPA2","S100A2","SFTPA1","NAPSA","CEACAM6","HOPX","KRT7","COL1A1","KRT19",
           "SFTA2","MDK","TM4SF1")

TPM_RiskScore <- TPM.norm[,genes]
TPM_RiskScore<-TPM_RiskScore %>%
  mutate(RiskScore=-0.1321*SFTPA2+(1.298e-05)*S100A2+0.1661*SFTPA1-0.1862*NAPSA+
           (2.932e-02)*CEACAM6-(2.593e-02)*HOPX+(4.562e-02)*KRT7+0.1063*COL1A1-
           (6.256e-02)*KRT19+(6.271e-02)*SFTA2+(3.384e-02)*MDK+0.1908*TM4SF1) %>%
  mutate(risk_group=ifelse(RiskScore>=median(RiskScore),'HRisk','LRisk'))

surv <- surv[rownames(TPM_RiskScore),]
survival_risk <- cbind(TPM_RiskScore[,c("RiskScore","risk_group")],surv)

fit<- survfit(Surv(round(futime/30.5,4), fustat) ~ risk_group, data = survival_risk,type = "kaplan-meier", error = "greenwood", conf.type = "plain", na.action = na.exclude)

# names(fit$strata) <- gsub("Subtype=", "", names(fit$strata))
ggsurv <- ggsurvplot(fit, data = survival_risk,
           surv.median.line = "h", # Add medians survival
           xlim = c(0, 120),
           break.time.by = 30,
           font.tickslab = c(12),
           font.x = c(14),
           font.y = c(14),

           # Change legends: title & labels
           legend.title = "TCGA-LUAD",
           legend.labs = c("High", "Low"),
           font.legend = c(12), 
           # Add p-value and tervals
           pval = TRUE,
           conf.int = FALSE,
           # Add risk table
           risk.table = TRUE,
           tables.height = 0.2,
           risk.table.col = "strata",
           risk.table.y.text = FALSE,

           # tables.theme = theme_cleantable(),
           # Color palettes. Use custom color: c("#E7B800", "#2E9FDF"),
           # or brewer color (e.g.: "Dark2"), or ggsci color (e.g.: "jco")
           palette = c("#E7B800", "#2E9FDF"),
           ggtheme = theme_classic() # Change ggplot2 theme
)
ggsurv$table <- ggsurv$table + labs(y = NULL) +
  theme(axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 14))
ggsurv

```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =5, fig.height =6}
# names(fit$strata) <- gsub("Subtype=", "", names(fit$strata))
ggsurv <- ggsurvplot(fit, data = survival_risk,
           surv.median.line = "h", # Add medians survival
           xlim = c(0, 120),
           break.time.by = 30,
           font.tickslab = c(12),
           font.x = c(14),
           font.y = c(14),

           # Change legends: title & labels
           legend.title = "",
           legend.labs = c("High", "Low"),
           font.legend = c(12), 
           # Add p-value and tervals
           pval = TRUE,
           conf.int = TRUE,
           # Add risk table
           risk.table = TRUE,
           tables.height = 0.2,
           risk.table.col = "strata",
           risk.table.y.text = FALSE,

           # tables.theme = theme_cleantable(),
           # Color palettes. Use custom color: c("#E7B800", "#2E9FDF"),
           # or brewer color (e.g.: "Dark2"), or ggsci color (e.g.: "jco")
           palette = "simpsons",
           # palette = c("#E7B800", "#2E9FDF"),
           ggtheme = theme_classic() # Change ggplot2 theme
)
ggsurv$table <- ggsurv$table + labs(y = NULL) +
  theme(axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 14))
ggsurv

pdf("./Results/Downstream_ggsurv.pdf",width=6,height=7)
ggsurv
dev.off()
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =6, fig.height =7}
survival_risk$Gender <- as.factor(ifelse(survival_risk$Gender %in% c("FEMALE"),"female","male"))
survival_risk$Stage <- as.factor(ifelse(survival_risk$pStage %in% c("Stage I","Stage IA","Stage IB"),"I",
                                        ifelse(survival_risk$pStage %in% c("Stage II","Stage IIA","Stage IIB"),"II",
                                               ifelse(survival_risk$pStage %in% c("Stage IIIA","Stage IIIB"),"III",
                                                      ifelse(survival_risk$pStage %in% c("Stage IV"),"IV",NA)))))
survival_risk$Race <- as.factor(ifelse(survival_risk$Race %in% c("ASIAN","BLACK OR AFRICAN AMERICAN","WHITE"),survival_risk$Race,"Others"))
survival_risk$Signature <-survival_risk$RiskScore

model <- coxph(Surv(futime, fustat) ~ Age + Gender + Stage + Signature, data=survival_risk)
m = summary(model)

p = ifelse(
  m$coefficients[, 5] < 0.001,
  "<0.001 ***",
  ifelse(
    m$coefficients[, 5] < 0.01,
    "<0.01  **",
    ifelse(
      m$coefficients[, 5] < 0.05,
      paste(round(m$coefficients[, 5], 3), " *"),
      round(m$coefficients[, 5], 3)
    )
  )
)

dat2 = as.data.frame(round(m$conf.int[, c(1, 3, 4)], 2))
dat2 = tibble::rownames_to_column(dat2, var = " ")
colnames(dat2)[2:4] = c("HR", "lower", "upper")

dat2$HR2 = paste0(dat2[, 2], " (", dat2[, 3], "-", dat2[, 4], ")")
dat2$P.Value = p

dat2$` ` = str_remove(dat2$` `, "Gender|Stage")
ins = function(x) {
  c(x, rep(NA, ncol(dat2) - 1))
}
dat2 = rbind(
  c(" ", NA, NA, NA, "HR", "P.Value"),
  dat2[1, ],
  ins("Gender"),
  ins("female"),
  dat2[2, ],
  ins("Stage"),
  ins("I"),
  dat2[3:nrow(dat2), ]
)
for(i in 2:4) {
  dat2[, i] = as.numeric(dat2[, i])
}

forestplot(
  dat2[, c(1, 5, 6)],
  mean = dat2[, 2],
  lower = dat2[, 3],
  upper = dat2[, 4],
  zero = 1,
  boxsize = 0.2,
  col = fpColors(box = '#1075BB', lines = 'black', zero = 'grey'),
  lty.ci = "solid",
  graph.pos = 2,
  #xticks = F,
  is.summary = c(T, F, T, F, F, T,F,F,F,F,F),
  align = "l",
  hrzl_lines = list(
    "2" = gpar(lty=1)),
  colgap = unit(5, 'mm')
)
```

#### GSE31210

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =5, fig.height =6}
load("GSE31210.Rdata")

TPM.norm <- log2(GSE31210 + 1)
TPM.norm <- data.frame(t(na.omit(TPM.norm[genes,])))
setdiff(genes,colnames(TPM.norm))
TPM_RiskScore <- TPM.norm %>%
  mutate(SFTPA2=0)
TPM_RiskScore<-TPM_RiskScore %>%
  mutate(RiskScore=-0.1321*SFTPA2+(1.298e-05)*S100A2+0.1661*SFTPA1-0.1862*NAPSA+
           (2.932e-02)*CEACAM6-(2.593e-02)*HOPX+(4.562e-02)*KRT7+0.1063*COL1A1-
           (6.256e-02)*KRT19+(6.271e-02)*SFTA2+(3.384e-02)*MDK+0.1908*TM4SF1) %>%
  mutate(risk_group=ifelse(RiskScore>=median(RiskScore),'HRisk','LRisk'))

clin <- clin[rownames(TPM_RiskScore),]
survival_risk <- cbind(TPM_RiskScore[,c("RiskScore","risk_group")],clin)
survival_risk$futime <- as.numeric(survival_risk$futime)

fit<- survfit(Surv(round(futime/30.5,4), fustat) ~ risk_group, data = survival_risk,type = "kaplan-meier", error = "greenwood", conf.type = "plain", na.action = na.exclude)
# names(fit$strata) <- gsub("Subtype=", "", names(fit$strata))
ggsurv <- ggsurvplot(fit, data = survival_risk,
           surv.median.line = "h", # Add medians survival
           xlim = c(0, 120),
           break.time.by = 30,
           font.tickslab = c(12),
           font.x = c(14),
           font.y = c(14),

           # Change legends: title & labels
           legend.title = "GSE31210",
           legend.labs = c("HRisk", "LRisk"),
           font.legend = c(12), 
           # Add p-value and tervals
           pval = TRUE,
           conf.int = FALSE,
           # Add risk table
           risk.table = TRUE,
           tables.height = 0.2,
           risk.table.col = "strata",
           risk.table.y.text = FALSE,

           # tables.theme = theme_cleantable(),
           # Color palettes. Use custom color: c("#E7B800", "#2E9FDF"),
           # or brewer color (e.g.: "Dark2"), or ggsci color (e.g.: "jco")
           palette = "simpsons",
           ggtheme = theme_classic() # Change ggplot2 theme
)
ggsurv$table <- ggsurv$table + labs(y = NULL) +
  theme(axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 14))
ggsurv

pdf("./Results/GSE31210_KM_plot.pdf",width=5,height=6)
ggsurv
dev.off()
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =6, fig.height =6}
survival_risk$Age <- as.numeric(survival_risk$`age (years):ch1`)
survival_risk$Gender <- as.factor(survival_risk$`gender:ch1`)
survival_risk$Stage <- as.factor(survival_risk$`pstage iorii:ch1`)
survival_risk$Smoke <- as.factor(survival_risk$`smoking status:ch1`)
survival_risk$Signature <-survival_risk$RiskScore*2
model <- coxph(Surv(futime, fustat) ~ Age + Gender + Stage + Signature, data=survival_risk)
m = summary(model)

#p值改一下格式，加上显著性
p = ifelse(
  m$coefficients[, 5] < 0.001,
  "<0.001 ***",
  ifelse(
    m$coefficients[, 5] < 0.01,
    "<0.01  **",
    ifelse(
      m$coefficients[, 5] < 0.05,
      paste(round(m$coefficients[, 5], 3), " *"),
      round(m$coefficients[, 5], 3)
    )
  )
)
#HR和它的置信区间
dat2 = as.data.frame(round(m$conf.int[, c(1, 3, 4)], 2))
dat2 = tibble::rownames_to_column(dat2, var = " ")
colnames(dat2)[2:4] = c("HR", "lower", "upper")
#需要在图上显示的HR文字和p值
dat2$HR2 = paste0(dat2[, 2], " (", dat2[, 3], "-", dat2[, 4], ")")
dat2$P.Value = p

dat2$` ` = str_remove(dat2$` `, "Gender|Stage")
ins = function(x) {
  c(x, rep(NA, ncol(dat2) - 1))
}
dat2 = rbind(
  c(" ", NA, NA, NA, "HR", "P.Value"),
  dat2[1, ],
  ins("Gender"),
  ins("female"),
  dat2[2, ],
  ins("Stage"),
  ins("I"),
  dat2[3:nrow(dat2), ]
)
for(i in 2:4) {
  dat2[, i] = as.numeric(dat2[, i])
}

p <- forestplot(
  dat2[, c(1, 5, 6)],
  mean = dat2[, 2],
  lower = dat2[, 3],
  upper = dat2[, 4],
  zero = 1,
  boxsize = 0.2,
  col = fpColors(box = '#1075BB', lines = 'black', zero = 'grey'),
  lty.ci = "solid",
  graph.pos = 2,
  title="GSE31210",
  #xticks = F,
  is.summary = c(T, F, T, F, F, T,F,F,F),
  align = "l",
  hrzl_lines = list(
    "2" = gpar(lty=1)),
  colgap = unit(5, 'mm')
)
p

pdf("./Results/GSE31210_forestplot.pdf",width=6,height=5)
p
dev.off()
```

#### GSE11969

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =5, fig.height =6}
load("GSE11969.Rdata")

TPM.norm <- log2(GSE11969 + 1)
TPM.norm <- data.frame(t(na.omit(TPM.norm[genes,])))
setdiff(genes,colnames(TPM.norm))
TPM_RiskScore <- TPM.norm %>%
  mutate(HOPX=0,SFTA2=0)
TPM_RiskScore<-TPM_RiskScore %>%
  mutate(RiskScore=-0.1321*SFTPA2+(1.298e-05)*S100A2+0.1661*SFTPA1-0.1862*NAPSA+
           (2.932e-02)*CEACAM6-(2.593e-02)*HOPX+(4.562e-02)*KRT7+0.1063*COL1A1-
           (6.256e-02)*KRT19+(6.271e-02)*SFTA2+(3.384e-02)*MDK+0.1908*TM4SF1) %>%
  mutate(risk_group=ifelse(RiskScore>=median(RiskScore),'HRisk','LRisk'))

clin <- clin[rownames(TPM_RiskScore),c(52:85)]
survival_risk <- cbind(TPM_RiskScore[,c("RiskScore","risk_group")],clin)
survival_risk$futime <- as.numeric(survival_risk$futime)

fit<- survfit(Surv(round(futime/30.5,4), fustat) ~ risk_group, data = survival_risk,type = "kaplan-meier", error = "greenwood", conf.type = "plain", na.action = na.exclude)
# names(fit$strata) <- gsub("Subtype=", "", names(fit$strata))
ggsurv <- ggsurvplot(fit, data = survival_risk,
           surv.median.line = "h", # Add medians survival
           xlim = c(0, 120),
           break.time.by = 30,
           font.tickslab = c(12),
           font.x = c(14),
           font.y = c(14),

           # Change legends: title & labels
           legend.title = "GSE11969",
           legend.labs = c("HRisk", "LRisk"),
           font.legend = c(12), 
           # Add p-value and tervals
           pval = TRUE,
           conf.int = FALSE,
           # Add risk table
           risk.table = TRUE,
           tables.height = 0.2,
           risk.table.col = "strata",
           risk.table.y.text = FALSE,

           # tables.theme = theme_cleantable(),
           # Color palettes. Use custom color: c("#E7B800", "#2E9FDF"),
           # or brewer color (e.g.: "Dark2"), or ggsci color (e.g.: "jco")
           palette = "simpsons",
           ggtheme = theme_classic() # Change ggplot2 theme
)
ggsurv$table <- ggsurv$table + labs(y = NULL) +
  theme(axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 14))
ggsurv

pdf("./Results/GSE11969_KM_plot.pdf",width=5,height=6)
ggsurv
dev.off()
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =6, fig.height =6}
survival_risk$Age <- as.numeric(survival_risk$`Age:ch1`)
survival_risk$Gender <- as.factor(ifelse(survival_risk$`Sex:ch1` %in% c("F"),"female","male"))
survival_risk$Stage <- as.factor(ifelse(survival_risk$`pStage:ch1` %in% c("IA","IB"),"I",
                                        ifelse(survival_risk$`pStage:ch1` %in% c("IIA","IIB"),"II","III")))
survival_risk$Signature <-survival_risk$RiskScore*5
model <- coxph(Surv(futime, fustat) ~ Age + Gender + Stage + Signature, data=survival_risk)
m = summary(model)

#p值改一下格式，加上显著性
p = ifelse(
  m$coefficients[, 5] < 0.001,
  "<0.001 ***",
  ifelse(
    m$coefficients[, 5] < 0.01,
    "<0.01  **",
    ifelse(
      m$coefficients[, 5] < 0.05,
      paste(round(m$coefficients[, 5], 3), " *"),
      round(m$coefficients[, 5], 3)
    )
  )
)

dat2 = as.data.frame(round(m$conf.int[, c(1, 3, 4)], 2))
dat2 = tibble::rownames_to_column(dat2, var = " ")
colnames(dat2)[2:4] = c("HR", "lower", "upper")

dat2$HR2 = paste0(dat2[, 2], " (", dat2[, 3], "-", dat2[, 4], ")")
dat2$P.Value = p

dat2$` ` = str_remove(dat2$` `, "Gender|Stage")
ins = function(x) {
  c(x, rep(NA, ncol(dat2) - 1))
}
dat2 = rbind(
  c(" ", NA, NA, NA, "HR", "P.Value"),
  dat2[1, ],
  ins("Gender"),
  ins("female"),
  dat2[2, ],
  ins("Stage"),
  ins("I"),
  dat2[3:nrow(dat2), ]
)
for(i in 2:4) {
  dat2[, i] = as.numeric(dat2[, i])
}

p <- forestplot(
  dat2[, c(1, 5, 6)],
  mean = dat2[, 2],
  lower = dat2[, 3],
  upper = dat2[, 4],
  zero = 1,
  boxsize = 0.2,
  col = fpColors(box = '#1075BB', lines = 'black', zero = 'grey'),
  lty.ci = "solid",
  graph.pos = 2,
  title="GSE11969",
  #xticks = F,
  is.summary = c(T, F, T, F, F, T,F,F,F,F),
  align = "l",
  hrzl_lines = list(
    "2" = gpar(lty=1)),
  colgap = unit(5, 'mm')
)
p

pdf("./Results/GSE11969_forestplot.pdf",width=6,height=5)
p
dev.off()
```

#### GSE72094

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =5, fig.height =6}
load("GSE72094.Rdata")

TPM.norm <- log2(GSE72094 + 1)
TPM.norm <- data.frame(t(na.omit(TPM.norm[genes,])))

TPM_RiskScore<-TPM.norm %>%
  mutate(RiskScore=-0.1321*SFTPA2+(1.298e-05)*S100A2+0.1661*SFTPA1-0.1862*NAPSA+
           (2.932e-02)*CEACAM6-(2.593e-02)*HOPX+(4.562e-02)*KRT7+0.1063*COL1A1-
           (6.256e-02)*KRT19+(6.271e-02)*SFTA2+(3.384e-02)*MDK+0.1908*TM4SF1) %>%
  mutate(risk_group=ifelse(RiskScore>=median(RiskScore),'HRisk','LRisk'))

clin <- clin[rownames(TPM_RiskScore),]
survival_risk <- cbind(TPM_RiskScore[,c("RiskScore","risk_group")],clin)

fit<- survfit(Surv(round(futime/30.5,4), fustat) ~ risk_group, data = survival_risk,type = "kaplan-meier", error = "greenwood", conf.type = "plain", na.action = na.exclude)
# names(fit$strata) <- gsub("Subtype=", "", names(fit$strata))
ggsurv <- ggsurvplot(fit, data = survival_risk,
           surv.median.line = "h", # Add medians survival
           # xlim = c(0, 120),
           break.time.by = 30,
           font.tickslab = c(12),
           font.x = c(14),
           font.y = c(14),

           # Change legends: title & labels
           legend.title = "GSE72094",
           legend.labs = c("HRisk", "LRisk"),
           font.legend = c(12), 
           # Add p-value and tervals
           pval = TRUE,
           conf.int = FALSE,
           # Add risk table
           risk.table = TRUE,
           tables.height = 0.2,
           risk.table.col = "strata",
           risk.table.y.text = FALSE,

           # tables.theme = theme_cleantable(),
           # Color palettes. Use custom color: c("#E7B800", "#2E9FDF"),
           # or brewer color (e.g.: "Dark2"), or ggsci color (e.g.: "jco")
           palette = "simpsons",
           ggtheme = theme_classic() # Change ggplot2 theme
)
ggsurv$table <- ggsurv$table + labs(y = NULL) +
  theme(axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 14))
ggsurv

pdf("./Results/GSE72094_KM_plot.pdf",width=5,height=6)
ggsurv
dev.off()
```

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.width =6, fig.height =7}
survival_risk$Age <- as.numeric(survival_risk$age_at_diagnosis)
survival_risk$Gender <- as.factor(ifelse(survival_risk$gender %in% c("F"),"female","male"))
survival_risk$Stage <- as.factor(ifelse(survival_risk$Stage %in% c("1","1A","1B"),"I",
                                        ifelse(survival_risk$Stage %in% c("2A","2B"),"II",
                                               ifelse(survival_risk$Stage %in% c("3A","3B"),"III","IV"))))
survival_risk$Signature <-survival_risk$RiskScore*7
model <- coxph(Surv(futime, fustat) ~ Age + Gender + Stage + Signature, data=survival_risk)
m = summary(model)

#p值改一下格式，加上显著性
p = ifelse(
  m$coefficients[, 5] < 0.001,
  "<0.001 ***",
  ifelse(
    m$coefficients[, 5] < 0.01,
    "<0.01  **",
    ifelse(
      m$coefficients[, 5] < 0.05,
      paste(round(m$coefficients[, 5], 3), " *"),
      round(m$coefficients[, 5], 3)
    )
  )
)
#HR和它的置信区间
dat2 = as.data.frame(round(m$conf.int[, c(1, 3, 4)], 2))
dat2 = tibble::rownames_to_column(dat2, var = " ")
colnames(dat2)[2:4] = c("HR", "lower", "upper")
#需要在图上显示的HR文字和p值
dat2$HR2 = paste0(dat2[, 2], " (", dat2[, 3], "-", dat2[, 4], ")")
dat2$P.Value = p

dat2$` ` = str_remove(dat2$` `, "Gender|Stage")
ins = function(x) {
  c(x, rep(NA, ncol(dat2) - 1))
}
dat2 = rbind(
  c(" ", NA, NA, NA, "HR", "P.Value"),
  dat2[1, ],
  ins("Gender"),
  ins("female"),
  dat2[2, ],
  ins("Stage"),
  ins("I"),
  dat2[3:nrow(dat2), ]
)
for(i in 2:4) {
  dat2[, i] = as.numeric(dat2[, i])
}

p <- forestplot(
  dat2[, c(1, 5, 6)],
  mean = dat2[, 2],
  lower = dat2[, 3],
  upper = dat2[, 4],
  zero = 1,
  boxsize = 0.2,
  col = fpColors(box = '#1075BB', lines = 'black', zero = 'grey'),
  lty.ci = "solid",
  graph.pos = 2,
  title="GSE72094",
  #xticks = F,
  is.summary = c(T, F, T, F, F, T,F,F,F,F,F),
  align = "l",
  hrzl_lines = list(
    "2" = gpar(lty=1)),
  colgap = unit(5, 'mm')
)
p

pdf("./Results/GSE72094_forestplot.pdf",width=6,height=5)
p
dev.off()
```

### GSEA by using KEGG genesets

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 4, fig.width = 9}
library(fgsea)
library(msigdbr)
library(openxlsx)

mdb_c2 <- msigdbr(species = "Homo sapiens", category = "C2")
mdb_kegg = mdb_c2 [grep("^KEGG",mdb_c2 $gs_name),]
fgsea_sets = mdb_kegg %>% split(x = .$gene_symbol, f = .$gs_name)

cluster <- c("Alive","Died")
ls <- list()
for (i in cluster) {
  makers <- SC.markers %>% filter(cluster==i)
  cluster.genes<- makers %>% arrange(desc(avg_log2FC)) %>% dplyr::select(gene,avg_log2FC)
  ranks<- deframe(cluster.genes)
  fgseaRes<- fgsea(fgsea_sets, stats = ranks)
  fgseaRes<- fgseaRes %>% arrange(desc(NES))
  ls <- c(ls,list(fgseaRes))
}
names(ls)[1] <- "Alive"
names(ls)[2] <- "Died"


ls1 <- ls[["Died"]]
ls1$pathway <- gsub('[_]', ' ', ls1$pathway)
ls1$pathway <- str_replace(ls1$pathway, "KEGG ","")
ls1$pathway <- str_replace(ls1$pathway, "CYTOKINE CYTOKINE ","CYTOKINE-CYTOKINE")

tmp <- ls1 %>% as_tibble() %>% 
  arrange(desc(NES)) %>% 
  filter(pval < 0.05) %>% 
  filter(NES>0) %>% head(n= 20)

library(RColorBrewer)
p <- ggplot(tmp, aes(reorder(pathway, NES), NES,fill= pval)) +
  geom_bar(stat ="identity",width = 0.7,position = position_dodge(width=0.8))+ 
#  geom_col(aes()) +
  # scale_fill_gradient(low = "#156077", high = "#f46f20")+
  # scale_fill_continuous(low='#225EA8', high='#C7E9B4') +
  scale_fill_distiller(palette = "YlGnBu",name="P.Value") +
  coord_flip() +
  labs(x="")+
  theme_bw()+
  # theme_classic()+
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14), 
        axis.title.y = element_text(size = 14),
        legend.text = element_text(size = 12))
p

pdf(file.path("./Results/KEGG Pathway Enrichment(Died).pdf"),width = 9,height = 4)
p
dev.off()
```

### GSEA by using HALLMARK genesets

```{r warning=FALSE, echo=FALSE, message=FALSE, fig.height = 8, fig.width = 8}
library(clusterProfiler)
library(enrichplot)

MSigDB=read.gmt("h.all.v2023.1.Hs.symbols.gmt")
MSigDB$term <- str_replace(MSigDB$term, "HALLMARK_","")
MSigDB$term <- str_replace_all(MSigDB$term, "_"," ")

data <- biomarkers
data$log2fc <- ifelse(data$cluster %in% c("Alive"),-data$avg_log2FC,data$avg_log2FC)
geneList <- data$log2fc
names(geneList) <- data$gene
geneList <- sort(geneList,decreasing = T)
GSEA.data <- GSEA(geneList = geneList,TERM2GENE=MSigDB,seed = T,verbose=F,pvalueCutoff = 0.25)

tmp <- GSEA.data@result %>% 
  arrange(desc(NES)) %>% 
  filter(p.adjust < 0.05) %>% 
  filter(NES>0) %>% head(n= 25)

p <- ggplot(tmp, aes(reorder(ID, NES), NES,fill= p.adjust)) +
  geom_bar(stat ="identity",width = 0.7,position = position_dodge(width=0.8))+ 
  scale_fill_distiller(palette = "YlGnBu",name="P.Adjust") +
  coord_flip() +
  labs(x="")+
  theme_bw()+
  # theme_classic()+
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14), 
        axis.title.y = element_text(size = 14),
        legend.text = element_text(size = 12))
p

pdf(file.path("./Results/GSEA enrichplot(Died).pdf"),width = 8,height = 8)
p
dev.off()
```
